### PR TITLE
allow setting service name from auto instrumentation context

### PR DESCRIPTION
**Pull request by pdelewski**
Currently there is no way to set service name from auto instrumentation context for zipkin and newrelic exporters.
We should read that from `OTEL_SERVICE_NAME` or any other configuration settings IMO.

### CODE REVIEW

**Comment by Nevay** [comment]
IMO the service name should not be a constructor argument and should be read in `::export()` from `SpanDataInterface::getResource()` instead.

**Comment by pdelewski** [comment]
> IMO the service name should not be a constructor argument and should be read in `::export()` from `SpanDataInterface::getResource()` instead.

Thx. That's the discussion I wanted to start opening this PR

**Comment by brettmc** [comment]
I think the spec agrees with Nevay here; the objective of providing the name in the exporter is eventually to pass it through as a service name, but that should indeed come from a resource attribute: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/README.md#service
Since not all exporters accept a name, it looks like we should remove the constructor argument totally (assuming we can instead get the service name out of resource attributes)

**Comment by pdelewski** [comment]
> IMO the service name should not be a constructor argument and should be read in `::export()` from `SpanDataInterface::getResource()` instead.

I'm currently looking into this code and have some concerns. `export` takes whole batch

**Comment by Nevay** [comment]
Yes, https://github.com/open-telemetry/opentelemetry-php/blob/587bc5aaf0b1c341391ac5843fa68e926852fe7a/src/Contrib/Zipkin/SpanConverter.php#L85 should be changed to something like
```php
'serviceName' => $span->getResource()->getAttributes()->get(ResourceAttributes::SERVICE_NAME),
```

**Comment by pdelewski** [comment]
> Yes,
> 
> https://github.com/open-telemetry/opentelemetry-php/blob/587bc5aaf0b1c341391ac5843fa68e926852fe7a/src/Contrib/Zipkin/SpanConverter.php#L85
> 
> should be changed to something like
> ```
> 'serviceName' => $span->getResource()->getAttributes()->get(ResourceAttributes::SERVICE_NAME),
> ```

Thx. That's exactly what I was thinking about

File: src/Contrib/Zipkin/SpanConverter.php [hunk 1]
```diff
@@ -78,11 +72,12 @@ private function convertSpan(SpanDataInterface $span): array
         $startTimestamp = TimeUtil::nanosToMicros($span->getStartEpochNanos());
         $endTimestamp = TimeUtil::nanosToMicros($span->getEndEpochNanos());
 
+        $serviceName =  $span->getResource()->getAttributes()->get(ResourceAttributes::SERVICE_NAME) ?? 'zipkin';
```
[the remark this answers is not in the document]
Comment (pdelewski) [reply src/Contrib/Zipkin/SpanConverter.php:75]: Yes, that's right, it's still work in progress

[the remark this answers is not in the document]
Comment (pdelewski) [reply src/Contrib/Zipkin/SpanConverter.php]: @Nevay fixed fallback

File: src/Contrib/Newrelic/Exporter.php [hunk 2]
```diff
@@ -23,25 +23,23 @@
  */
 class Exporter implements SpanExporterInterface
 {
+    public $name;
```
Comment (Nevay) [inline src/Contrib/Newrelic/Exporter.php:26]: Should be removed.

File: src/Contrib/Newrelic/SpanConverter.php [hunk 3]
```diff
@@ -40,12 +35,16 @@ private function convertSpan(SpanDataInterface $span): array
         $startTimestamp = TimeUtil::nanosToMillis($span->getStartEpochNanos());
         $endTimestamp = TimeUtil::nanosToMillis($span->getEndEpochNanos());
 
+        $serviceName =  $span->getResource()->getAttributes()->get(ResourceAttributes::SERVICE_NAME)
+                        ??
+                        ResourceInfoFactory::defaultResource()->getAttributes()->get(ResourceAttributes::SERVICE_NAME);
```
Comment (brettmc) [inline src/Contrib/Newrelic/SpanConverter.php:40]: depending on how many detectors are enabled (by default, all of them are), this is a bit of overhead we could avoid? Perhaps setting a default/fallback service name in the constructor?

Comment (pdelewski) [reply src/Contrib/Newrelic/SpanConverter.php:40]: Good point

File: src/Contrib/Newrelic/SpanConverter.php [hunk 4]
```diff
@@ -18,9 +20,9 @@ class SpanConverter implements SpanConverterInterface
 
     private string $serviceName;
```
Comment (brettmc) [inline src/Contrib/Newrelic/SpanConverter.php:21]: a better name might be `fallbackServiceName` or `defaultServiceName` ?

File: tests/Unit/Contrib/Zipkin/ZipkinSpanConverterTest.php [hunk 5]
```diff
@@ -61,14 +61,14 @@ public function test_should_convert_a_span_to_a_payload_for_zipkin(): void
...
-        $converter = new SpanConverter('test.name');
+        $converter = new SpanConverter();
         $row = $converter->convert([$span])[0];
 
         $this->assertSame($span->getContext()->getSpanId(), $row['id']);
         $this->assertSame($span->getContext()->getTraceId(), $row['traceId']);
         $this->assertSame('1000000000000000', $row['parentId']);
 
-        $this->assertSame('test.name', $row['localEndpoint']['serviceName']);
+        $this->assertSame('unknown_service', $row['localEndpoint']['serviceName']);
```
Comment (brettmc) [inline tests/Unit/Contrib/Zipkin/ZipkinSpanConverterTest.php:71]: you might need to merge in main branch, this is something I changed in a commit in the last day or so... `unknown_service:php`

Comment (pdelewski) [reply tests/Unit/Contrib/Zipkin/ZipkinSpanConverterTest.php:71]: yes, that's right. Will do that when this PR will be approved

**Review by brettmc — APPROVED** [review]

**Pull request merged**